addrmgr/addrmanager.go

```diff
@@ -900,86 +999,94 @@ func (a *AddrManager) Good(addr *wire.NetAddress) {
+	// If the address is already tried then return since it's already good.
+	// Otherwise, move it to a tried bucket. If the target tried bucket is full,
+	// then room will be made by evicting the oldest address in that bucket and
+	// moving it to a new bucket. If the psuedorandomly selected new bucket is
+	// full, then swap the addresses' positions between tried and new.
 	if ka.tried {
-		return
+		return nil
 	}
 
-	// ok, need to move it to tried.
-
 	// remove from all new buckets.
 	// record one of the buckets in question and call it the `first'
 	addrKey := NetAddressKey(addr)
-	oldBucket := -1
+	addrNewAvailableIndex := -1
 	for i := range a.addrNew {
 		// we check for existence so we can record the first one
 		if _, ok := a.addrNew[i][addrKey]; ok {
 			delete(a.addrNew[i], addrKey)
 			a.addrChanged = true
 			ka.refs--
-			if oldBucket == -1 {
-				oldBucket = i
+			if addrNewAvailableIndex == -1 {
+				addrNewAvailableIndex = i
 			}
 		}
 	}
 	a.nNew--
 
-	if oldBucket == -1 {
-		// What? wasn't in a bucket after all.... Panic?
-		return
+	if addrNewAvailableIndex == -1 {
+		str := fmt.Sprintf("%s is not marked as a new address", ipString(addr))
+		return makeError(ErrAddressNotFound, str)
 	}
 
 	bucket := a.getTriedBucket(ka.na)
 
-	// Room in this tried bucket?
-	if len(a.addrTried[bucket]) < triedBucketSize {
+	// If this tried bucket has enough capacity for another address,
+	// add the address to the bucket and flag it as tried.
+	if len(a.addrTried[bucket]) < a.triedBucketSize {
 		ka.tried = true
 		a.addrTried[bucket] = append(a.addrTried[bucket], ka)
 		a.addrChanged = true
 		a.nTried++
-		return
+		return nil
 	}
 
-	// No room, we have to evict something else.
-	triedIdx := a.pickTried(bucket)
-	rmka := a.addrTried[bucket][triedIdx]
+	// Since the tried bucket is at capacity, evict the oldest address
+	// in the tried bucket and move it to a new bucket.
+	oldestTriedIndex := a.getOldestAddressIndex(bucket)
+	rmka := a.addrTried[bucket][oldestTriedIndex]
 
 	// First bucket it would have been put in.
 	newBucket := a.getNewBucket(rmka.na, rmka.srcAddr)
 
-	// If no room in the original bucket, we put it in a bucket we just
-	// freed up a space in.
+	// If there is no room in the psuedorandomly selected new bucket,
+	// then reuse the new bucket that the newly tried address was removed from.
 	if len(a.addrNew[newBucket]) >= newBucketSize {
-		newBucket = oldBucket
+		newBucket = addrNewAvailableIndex
 	}
 
-	// replace with ka in list.
+	// Replace oldest tried address in bucket with ka.
 	ka.tried = true
-	a.addrTried[bucket][triedIdx] = ka
+	a.addrTried[bucket][oldestTriedIndex] = ka
 
 	rmka.tried = false
 	rmka.refs++
 
-	// We don't touch a.nTried here since the number of tried stays the same
-	// but we decremented a.nNew above, raise it again since we're putting
-	// something back.
+	// The total number of tried addresses a.nTried is not modified here since
```

Comment:
```suggestion
	// The total number of tried addresses is not modified here since
```

It's best practice to avoid referencing variable names in comments since they often change over time and the comments don't get updated.  Instead, they should describe overall intent.

Reply:
That makes sense, thanks for the tip. Updated comments. Some of the tests were also referencing variable names and those were updated as well.